src/game/behaviors/fish.inc.c

```diff
@@ -1,163 +1,299 @@
-        sp18 *= 127;
+        o->oFishRandomDistance2 = RandomFloat() * 300.0f;
+        o->oFishRandomSpeed = RandomFloat() * 1024.0f + 1024.0f;
+        o->oFishRandomVel = RandomFloat() * 4.0f + 8.0f + 5.0f;
+        if (o->oDistanceToMario < 600.0f) {
+            distance = 1;
+        } else {
+            distance = (s32)(1.0 / (o->oDistanceToMario / 600.0));
+        }
+        distance *= 127;
         PlaySound2(SOUND_GENERAL_MOVING_WATER);
     }
-    if (o->oTimer < LEVEL_SA)
+    // Enable fish animation. Parameters differ if oTimer < SA.
+    if (o->oTimer < 20) {
         func_8029ED98(0, 4.0f);
-    else
+    } else {
         func_8029ED98(0, 1.0f);
-    if (o->oForwardVel < o->oFishGroupUnk108)
+    }
+    // Set forward velocity
+    if (o->oForwardVel < o->oFishRandomVel) {
         o->oForwardVel = o->oForwardVel + 0.5;
-    o->oFishGroupUnkF8 = gMarioObject->oPosY + o->oFishGroupUnkFC;
-    obj_rotate_yaw_toward(o->oAngleToMario + 0x8000, o->oFishGroupUnk100);
-    if (o->oPosY < o->oFishGroupUnkF4 - 50.0f) {
-        if (sp1C < 0.0f)
-            sp1C = 0.0f - sp1C;
-        if (sp1C < 500.0f)
-            func_802BF59C(2);
-        else
-            func_802BF59C(4);
+    }
+    o->oFishPosY = gMarioObject->oPosY + o->oFishRandomOffset;
+    // Rotate fish towards mario
+    obj_rotate_yaw_toward(o->oAngleToMario + 0x8000, o->oFishRandomSpeed);
+    // If fish groups are too close, call fish_regroup()
+    if (o->oPosY < o->oFishWaterLevel - 50.0f) {
+        if (fishY < 0.0f) {
+            fishY = 0.0f - fishY;
+        }
+        if (fishY < 500.0f) {
+            fish_regroup(2);
+        } else {
+            fish_regroup(4);
+        }
     } else {
-        o->oPosY = o->oFishGroupUnkF4 - 50.0f;
-        if (sp1C > 300.0f)
+        o->oPosY = o->oFishWaterLevel - 50.0f;
+        if (fishY > 300.0f) {
             o->oPosY -= 1.0f;
+        }
+    }
+    // If distance to mario is too great, then set fish to active.
+    if (o->oDistanceToMario > o->oFishRandomDistance2 + 500.0f) {
+        o->oAction = FISH_ACTIVE;
     }
-    if (o->oDistanceToMario > o->oFishGroupUnk110 + 500.0f)
-        o->oAction = 1;
 }
-
-void ActionFishGroup0(void) {
+/**
+ * Animate fish and alter scaling at random
+ */
+void fish_group_act_animate(void) {
     func_8029ED98(0, 1.0f);
     o->header.gfx.unk38.animFrame = (s16)(RandomFloat() * 28.0f);
-    o->oFishGroupUnk10C = RandomFloat() * 300.0f;
+    o->oFishRandomYDistance = RandomFloat() * 300.0f;
     obj_scale(RandomFloat() * 0.4 + 0.8);
-    o->oAction = 1;
+    o->oAction = FISH_ACTIVE;
 }
 
-void (*sFishGroupActions[])(void) = { ActionFishGroup0, ActionFishGroup1, ActionFishGroup2 };
+/**
+ * Array of methods
+ */
+void (*sFishGroupActions[])(void) = { 
+    fish_group_act_animate, 
+    fish_group_act_rotation, 
+    fish_group_act_move
+};
 
-void bhv_fish_group_2_loop(void) // TODO rename
+/**
+ * Main loop for fish
+ */
+void bhv_fish_group_2_loop(void)
```

Comment:
Wait. wait. wait.

`bhv_fish_group_2_loop` is the main loop for fish, instead of `bhv_fish_loop`???
What is the _2 doing in that name?
If it's not the actual fish behavior, why does the fish *group* use object fields with prefix oFish, and have action macros like `FISH_RESPAWN` instead of `FISH_GROUP_RESPAWN`?
This is very confusing.

Reply:
I think the *reality* might be something similar to birds, where there's 1 "leader" bird that's loaded from the start of the level and spawns in all the other birds, but otherwise acts the same as a regular bird. Is this a correct guess?

Reply:
I will consider this next time I reading through the file.
I have not changed the names of BHV_ methods because they are predetermined in another file. Of course I can change that one too. Just for now I haven't. Should I change it?

Reply:
Yes of course! Otherwise your documentation and the function's names conflict with each other.